### PR TITLE
[WIP] Merge in code as per v2Plan.md

### PR DESCRIPTION
**Pull request by bobbydeveaux**

### CODE REVIEW

**Comment by kradalby** [comment]
Status of this pull request:

We (ESA) currently use this internally and it seem to work pretty well.

The restructuring we have done pleases us a lot more than the old structure (take a look at the new documentation).

There is a few things to note:
- The API level has been set to vCloud Director 8.0
- This depends on our/my fork of [govcloudair](https://github.com/kradalby/govcloudair/) which has a lot of rewrites and breaks a lot of backwards compat.
- The test suit is not run as I currently just have tests that run in our environment.
- Changes dependency management to dep

I have been working on other things for a while, but we have a lot of interest in getting these changes upstream, so I would like to get feedback on how we can do that and more people to test.

This should also solve the following issues:
#9 (most of)
#27 
#28 
#50 
#52 
#62 
#75 

And I believe also the following issues, but I need testers:
#7 
#16
#26  
#63 
#69 
#81 


I can devote a certain amount of time to the project, and fix bugs if the arise, but doing things that goes outside our scope of usage is hard to get done at the moment.

@bobbydeveaux  @dn1s

**Comment by dn1s** [comment]
So I finally got to test it...
There were two things which did not work for me.
in `vendor/github.com/kradalby/govcloudair/types/v56/types.go` I had to reorder elements in struct since
vcd api is expecting them in this order:
```
type VMCapabilities struct {
    HREF                string   `xml:"href,attr,omitempty"`
    Type                string   `xml:"type,attr,omitempty"`
    Link                LinkList `xml:"Link,omitempty"`
    MemoryHotAddEnabled bool     `xml:"MemoryHotAddEnabled,omitempty"`
    CPUHotAddEnabled    bool     `xml:"CpuHotAddEnabled,omitempty"`
}
```
and in `vcd/helper_vm.go`:
```
diff --git a/vcd/helper_vm.go b/vcd/helper_vm.go
index 4cce234..58ed20f 100644
--- a/vcd/helper_vm.go
+++ b/vcd/helper_vm.go
@@ -59,7 +59,7 @@ func composeSourceItem(d *schema.ResourceData, meta interface{}) (*types.Sourced
                InstantiationParams: &types.InstantiationParams{
                        NetworkConnectionSection: vm.VM.NetworkConnectionSection,
                        // GuestCustomizationSection: vm.VM.GuestCustomizationSection,
-                       OVFVirtualHardwareSection: vm.VM.VirtualHardwareSection.ConvertToOVF(),
+                       // OVFVirtualHardwareSection: vm.VM.VirtualHardwareSection.ConvertToOVF(),
                },
        }
 
```
This change is needed because InstantiationParams can't have disks in it and CPU and Memory settings are ignored in a RecomposeVM run. Since we change all settings later anyways I patched it out locally for me. After those patches creating a vm with multiple vms with multiple network adapters worked as well as scaling those vms up and down.

Please check if these changes are breaking anything on your side.

Kind regards,

Dominik

**Comment by spengilley** [comment]
I'm in need of multiple networks on a VM. Is this going to be merged in at any point?

**Comment by dn1s** [comment]
I'm right now developing this feature you can check out my code here:
https://github.com/dn1s/terraform-provider-vcd/tree/add_multi_nics_support
Note this is not yet merged neither to https://github.com/vmware/go-vcloud-director nor to this repo!

**Comment by spengilley** [comment]
Thanks, man :)

**Comment by lvirbalas** [comment]
@dn1s when do you plan to post a PR? I see you're working in the VM resource which is quite an active one, so you may end up with conflicts if you hold on to the code long enough. I suggest you update your fork from master often.

One related heads up is that @vbauzysvmware is working on adding **vApp level network support** as part of #155

**Comment by lvirbalas** [comment]
Spring cleaning heads up. We were not here when this PR was opened, but now it's impossible to merge it due to its sheer size and changes that happened in between. Overall, PRs need to be as small as possible and incorporate one feature at a time.

However, as opposed to just closing it we are planning to review the code and post new issues for the features and fixes that are possible to port over to current HEAD. Then we can cherry-pick some code and manually port what is useful.

**Comment by bobbydeveaux** [comment]
Thanks @lvirbalas - yes it is a shame to throw this code away. Shame we didn't break this down into small PR's for sure!! Thanks for your great work on improving it!

**Comment by lvirbalas** [comment]
Thanks @bobbydeveaux for understanding.

@vbauzysvmware will bear the torch of this quantisation effort!

File: vendor/github.com/ukcloud/govcloudair/vm.go [hunk 1]
```diff
@@ -522,3 +290,74 @@ func (v *VM) GetMemoryCount() (int, error) {
 
 	return items[0].VirtualQuantity, nil
 }
+
+func (v *VM) correctAddressOnParentForNetworkHardware() error {
```
Comment (vbauzysvmware) [inline vendor/github.com/ukcloud/govcloudair/vm.go]: @kradalby  Could you explain what this function purpose? what issue it solves?

**Comment by vbauzysvmware** [comment]
As we agreed, code reviewed and  possible improvements already taken and integrated to current code:
https://github.com/vmware/go-vcloud-director/pull/179
https://github.com/vmware/go-vcloud-director/pull/182
https://github.com/terraform-providers/terraform-provider-vcd/issues/219

The biggest change in this PR is for vAPP and VM. It will be handled and issue for tracking is this: 
https://github.com/terraform-providers/terraform-provider-vcd/issues/217

**Pull request closed without merging**